### PR TITLE
Lambada calc by Oleh Kravets

### PR DESCRIPTION
**Pull request by Ketch92**

### CODE REVIEW

File: src/main/java/core/basesyntax/Calculator.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        if (operation.equals(DIV)) {
+            if (second == 0) {
+                throw new ArithmeticException("Can't divide by 0");
+            }
+            return (double) first / second;
+        }
+        if (operation.equals(MUL)) {
+            return (double) first * second;
+        }
+        if (operation.equals(POW) || operation.equals(POW.toUpperCase())) {
```
Comment (Artificial-Friend) [inline src/main/java/core/basesyntax/Calculator.java:39]: If I do calculate(1, 2, POW) your program will invoke 5 unnecessary if() statements. Are you sure it is a correct way to do it?

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    private static final String ADD = "+";
+    private static final String SUB = "-";
+    private static final String DIV = "/";
+    private static final String MUL = "*";
+    private static final String POW = "p";
+    private static final String POW_UPPERCASE = "P";
+
+    @BeforeEach
+    void setUp() {
+        magic = new Calculator();
```
Comment (Artificial-Friend) [inline src/test/java/core/basesyntax/CalculatorTest.java:23]: If I would need to use your Calculator instance and start your program I would be lost in that naming. `magic` could be `MAX_VALUE`.

File: src/main/java/core/basesyntax/Calculator.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    private static final String MUL = "*";
+    private static final String POW = "p";
+
+    /*
+    + second adds to first
+    - second subtracts from fist
+    / first is divisor and second is divider
+    * second times first
+    p first in the power of second
+     */
```
Comment (Artificial-Friend) [inline src/main/java/core/basesyntax/Calculator.java:19]: Is it necessary? Also it says `- second subtracts from fist` but 28 line subtract first from second

**Review by Artificial-Friend — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Calculator.java (hunk 1 above)
Comment (Ketch92) [reply src/main/java/core/basesyntax/Calculator.java:39]: I totally agree that using a switch is better but I'm a bit confused with "common mistakes" where is written to not use it.

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 2 above)
Comment (Ketch92) [reply src/test/java/core/basesyntax/CalculatorTest.java:23]: True, I was kind of messing up a bit. But for sure it isn't proper naming

File: src/main/java/core/basesyntax/Calculator.java (hunk 3 above)
Comment (Ketch92) [reply src/main/java/core/basesyntax/Calculator.java:19]: It was rather for me)

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+            }
+        }
+    }
+
+    @Test
+    void random_input_division() {
+        int a;
+        int b;
+        for (int i = 0; i < 1000; i++) {
+            a = new Random().nextInt();
```
Comment (Artificial-Friend) [inline src/test/java/core/basesyntax/CalculatorTest.java:226]: While I understand it, Bohdan said that any usage of random in tests is a bad practice. I hope mentors will answer if it is. From my perspective this is very useful testing method.
I also I am a little bit confused is it really that necessary to create constants for operators, because your method using '*' and you giving '*' to the method but in code i see `MUL`. Then again this question is more to mentors, I am not sure what option will be better.

**Review by Artificial-Friend — APPROVED** [review]

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 4 above)
Comment (dmytro-hryhoruk) [reply src/test/java/core/basesyntax/CalculatorTest.java:226]: 1) It's very hard to debug random tests and it's always impossible to reproduce them.
Also, It's running your code always with different random parameters, so it may be the situation when you have executed your tests once and they have passed, but the second time they haven't.  There is no point in making 1000 similar test cases. It's better to make a few but with different behavior.
2) don't use MUl as a constant, write the full name of the operation

Comment (dmytro-hryhoruk) [reply src/test/java/core/basesyntax/CalculatorTest.java:226]: for now, you can leave the random test just keep in mind the information, also there is a good answer on StackOverflow 
https://stackoverflow.com/questions/3441686/what-are-the-downsides-using-random-values-in-unit-testing

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 4 above)
Comment (Ketch92) [reply src/test/java/core/basesyntax/CalculatorTest.java:226]: I appreciate your concerns on the topic of randomized tests, and I remember Bohdan's warnings on that. I added these randomized tests just because the expected output is clear and manageable to compare with the implemented method. For the additional task given by Bohdan (the elevator task) I also supposed to write the tests, but none of them is randomized as it's possible to get an unexpected result not fitting to the task at all.
Dmytro, thanks for the post.

File: pom.xml [hunk 5]
```diff
@@ -12,26 +12,28 @@
...
         <dependency>
             <groupId>org.junit.jupiter</groupId>
-            <artifactId>junit-jupiter</artifactId>
-            <version>${junit-jupiter.version}</version>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.7.0</version>
             <scope>test</scope>
         </dependency>
     </dependencies>
+
```
[the remark this answers is not in the document]
Comment (Ketch92) [reply pom.xml:30]: In the video for these lessons, Bohdan added this dependency. And right after cloning the repo I had troubles. And this change fixed it. Or was there another way to handle it?

File: src/main/java/core/basesyntax/Calculator.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    private static final String SUB = "-";
+    private static final String DIV = "/";
+    private static final String MUL = "*";
+    private static final String POW = "p";
+    private static final String POW_UPPERCASE = "P";
+
+    public double calculate(int first, int second, String operation) {
+        if (operation == null) {
+            throw new NullPointerException();
+        }
```
[the remark this answers is not in the document]
Comment (Ketch92) [reply src/main/java/core/basesyntax/Calculator.java:17]: Ok, removed)

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            b = new Random().nextInt();
+            if ((double) a / b != calculator.calculate(a, b, DIV)) {
+                Assertions.fail("Expected value "+ (double) a / b
+                        + "but was " + calculator.calculate(a, b, DIV));
+            }
+        }
+    }
+
+
+}
```
[the remark this answers is not in the document]
Comment (Ketch92) [reply src/test/java/core/basesyntax/CalculatorTest.java:236]: removed as well

**Review by dmytro-hryhoruk — APPROVED** [review]
